src/loader/loader.c

```diff
@@ -0,0 +1,302 @@
+#include <postgres.h>
+#include <pg_config.h>
+#include <access/xact.h>
+#include <commands/extension.h>
+#include <miscadmin.h>
+#include <utils/guc.h>
+#include <utils/inval.h>
+#include <parser/analyze.h>
+#include <nodes/print.h>
+
+#define EXTENSION_NAME "timescaledb"
+
+#include "../extension_utils.c"
+
+#define PG96 ((PG_VERSION_NUM >= 90600) && (PG_VERSION_NUM < 100000))
+#define PG10 ((PG_VERSION_NUM >= 100000) && (PG_VERSION_NUM < 110000))
+/*
+ * Some notes on design:
+ *
+ * We do not check for the installation of the extension upon loading the extension and instead rely on a hook for two reasons:
+ * 1) We probably can't
+ *	- The shared_preload_libraries is called in PostmasterMain which is way before InitPostgres is called.
+ *			(Note: This happens even before the fork of the backend) -- so we don't even know which database this is for.
+ *	-- This means we cannot query for the existance of the extension yet because the caches are initialized in InitPostgres.
+ * 2) We actually don't want to load the extension in two cases:
+ *	  a) We are upgrading the extension.
+ *	  b) We set the guc timescaledb.disable_load.
+ *
+ *
+ */
+
+#ifdef PG_MODULE_MAGIC
+PG_MODULE_MAGIC;
+#endif
+
+#define GUC_DISABLE_LOAD_NAME "timescaledb.disable_load"
+
+extern void _PG_init(void);
+extern void _PG_fini(void);
+
+/* was the versioned-extension loaded*/
+static bool loaded = false;
+
+/* GUC to disable the load */
+static bool guc_disable_load = false;
+
+/* This is the hook that existed before the loader was installed */
+static post_parse_analyze_hook_type prev_post_parse_analyze_hook;
+
+/* This is timescaleDB's versioned-extension's post_parse_analyze_hook */
+static post_parse_analyze_hook_type extension_post_parse_analyze_hook = NULL;
+
+static void inline extension_check(void);
+static void call_extension_post_parse_analyze_hook(ParseState *pstate,
+									   Query *query);
+
+
+static void
+inval_cache_callback(Datum arg, Oid relid)
+{
+	if (guc_disable_load)
+		return;
+	extension_check();
+}
+
+
+static bool
+drop_statement_drops_extension(DropStmt *stmt)
+{
+	if (stmt->removeType == OBJECT_EXTENSION)
+	{
+		if (list_length(stmt->objects) == 1)
+		{
+			char	   *ext_name;
+#if PG96
+			List	   *names = linitial(stmt->objects);
+
+			Assert(list_length(names) == 1);
+			ext_name = strVal(linitial(names));
+#elif PG10
+			void	   *name = linitial(stmt->objects);
+			ext_name = strVal(name);
+#endif
+			if (strcmp(ext_name, EXTENSION_NAME) == 0)
+				return true;
+		}
+	}
+	return false;
+}
+
+static bool
+load_utility_cmd(Node *utilityStmt)
+{
+	switch (nodeTag(utilityStmt))
+	{
+		case T_VariableSetStmt:
+			if (strcmp(((VariableSetStmt *) utilityStmt)->name, GUC_DISABLE_LOAD_NAME) == 0)
+				/* Do not load when setting the guc */
+				return false;
+			return true;
+		case T_AlterExtensionStmt:
+			if (strcmp(((AlterExtensionStmt *) utilityStmt)->extname, EXTENSION_NAME) != 0)
```

Comment:
This is not quite correct as you should return true if `strcmp(((AlterExtensionStmt *) utilityStmt)->extname, EXTENSION_NAME) != 0` and false otherwise. Left unchanged.